cmds/replay.c

```diff
@@ -321,6 +321,9 @@ static int print_flat_rstack(struct uftrace_data *handle, struct uftrace_task_re
 	if (!opts->libcall && sym && sym->type == ST_PLT_FUNC)
 		goto out;
 
+	if (opts->no_preempt_sched && rstack->addr == EVENT_ID_PERF_SCHED_OUT_PREEMPT)
+		goto out;
+
```

Comment:
Instead of checking this in commands, I think it should be handled in the fstack code.  Please see `fstack_check_opts()`.

Reply:
If I add or move the above if statement to fstack_check_opt,
The linux:sched-out (pre-empted) disappears as shown below, but the linux:sched-in is new.

```
# DURATION     TID     FUNCTION
   1.400 us [ 22028] | __monstartup();
   1.100 us [ 22028] | __cxa_atexit();
            [ 22028] | main() {
 209.500 us [ 22028] |   fork();
            [ 22028] |   wait() {
            [ 22028] |     /* linux:sched-out */
            [ 22031] |   } /* fork */
            [ 22031] |   a() {
            [ 22031] |     b() {
            [ 22031] |       c() {
  11.500 us [ 22031] |         /* linux:sched-in */
  11.100 us [ 22031] |         /* linux:sched-in */
  10.700 us [ 22031] |         /* linux:sched-in */
   3.100 us [ 22031] |         getpid();
   1.987  s [ 22031] |       } /* c */
   1.987  s [ 22031] |     } /* b */
   1.987  s [ 22031] |   } /* a */
   1.987  s [ 22031] | } /* main */
   1.988  s [ 22028] |     /* linux:sched-in */
   1.988  s [ 22028] |   } /* wait */
            [ 22028] |   a() {
            [ 22028] |     b() {
            [ 22028] |       c() {
  10.200 us [ 22028] |         /* linux:sched-in */
   3.800 us [ 22028] |         getpid();
   2.072  s [ 22028] |       } /* c */
   2.072  s [ 22028] |     } /* b */
   2.072  s [ 22028] |   } /* a */
   4.061  s [ 22028] | } /* main */
```

I think it's because the linux:sched-in is also recorded after the linux:sched-out (pre-empted), as shown in the log below when do uftrace replay --flat.

```
[13] !!! 22031: linux:sched-out (pre-empted) time (40562014989000)
[14] !!! 22031: linux:sched-in time (40562015000500)
[15] !!! 22031: linux:sched-out (pre-empted) time (40562895137400)
[16] !!! 22031: linux:sched-in time (40562895148500)
[17] !!! 22031: linux:sched-out (pre-empted) time (40563055016900)
[18] !!! 22031: linux:sched-in time (40563055027600)
```

But, if using uftrace replay or an uftrace graph, linux:sched-in is not showing, but only show linux:sched-out (pre-empted), and I think the reason is because linux:sched-in internally remove like the code below.

https://github.com/namhyung/uftrace/blob/aa24650f53f0899c8e635b580c13d089e1b73d1a/cmds/replay.c#L935-L955

So I thought it was right to filter in the command's print function after internal event processing is completed, so I wrote the above code to filter in the command's print function, and I tried to apply the filtering to fstack_check_options according to review, but I couldn't check if the linux:sched-out (pre-empted) event came out right before the linux:sched-in event. so is there a way to check the previous event in the uftrace_task_reader?

Reply:
Hmm.. maybe you need to add `sched_preempt_seen` to `uftrace_task_reader` and check it with `sched-in` event.